core/trino-grammar/src/main/antlr4/io/trino/grammar/sql/SqlBase.g4

```diff
@@ -198,11 +202,23 @@ statement
     ;
 
 rootQuery
-    : withFunction? query
+    : queryScoped? query
+    ;
+
+queryScoped
+    : WITH withFunction? withSession?
     ;
 
 withFunction
-    : WITH functionSpecification (',' functionSpecification)*
+    : functionSpecification (',' functionSpecification)*
+    ;
+
+withSession
+    : sessionSpecification (',' sessionSpecification)*
+    ;
+
+sessionSpecification
```

Comment:
The session is comprised of the set of property assignments, so a syntax that would be more appropriate is:

```
WITH SESSION
    property1 = value1,
    property2 = value2,
    ...
```

Reply:
This change makes grammar ambigous as:

```
namedQuery
    : name=identifier (columnAliases)? AS '(' query ')'
    ;
with
    : WITH RECURSIVE? namedQuery (',' namedQuery)*
    ;
query
    : with? queryNoWith
    ;
rootQuery
    : queryScoped? query
    ;

queryScoped
    : WITH withFunction? withSession?
    ;
```

which in turn makes this SQL invalid:
```
WITH  SESSION session_key = 'property', catalog.catalog_session_key = 'catalog_property', catalog.catalog_session_key2 = true
SELECT foo()
```

fails with:
```
io.trino.sql.parser.ParsingException: line 2:13: mismatched input '='. Expecting: '(', 'AS'
	at io.trino.sql.parser.ErrorHandler.syntaxError(ErrorHandler.java:108)
	at org.antlr.v4.runtime.ProxyErrorListener.syntaxError(ProxyErrorListener.java:41)
	at org.antlr.v4.runtime.Parser.notifyErrorListeners(Parser.java:544)
	at org.antlr.v4.runtime.DefaultErrorStrategy.reportInputMismatch(DefaultErrorStrategy.java:327)
	at org.antlr.v4.runtime.DefaultErrorStrategy.reportError(DefaultErrorStrategy.java:139)
	at io.trino.grammar.sql.SqlBaseParser.namedQuery(SqlBaseParser.java:8679)
```

Reply:
How can I make this non-ambigous @martint ?